site/frontend/src/pages/status/page.vue

```diff
@@ -19,84 +37,262 @@ function formatDuration(seconds: number): string {
-    }`;
-  } else if (reason["Master"] !== undefined && reason.Master.pr == 0) {
-    return "Master";
-  } else if (reason["Try"] !== undefined && reason.Try.pr) {
-    return `
-                Try for
-                <a href="https://github.com/rust-lang/rust/pull/${reason["Try"].pr}">
-                    #${reason["Try"].pr}
-                </a>`;
-  } else {
-    // Should never happen, but a reasonable fallback
-    return JSON.stringify(reason);
+function commitUrlAsHtml(sha: string): string {
+  return `<a href="https://github.com/rust-lang/rust/commit/${sha}">${sha.substring(
+    0,
+    13
+  )}</a>`;
+}
+
+function pullRequestUrlAsHtml(pr: number | null): string {
+  if (pr === null) {
+    return "";
   }
+  return `<a href="https://github.com/rust-lang/rust/pull/${pr}">#${pr}</a>`;
+}
+
+function formatCommitAsHtml(commit: Commit, reason: MissingReason): string {
+  const url = commitUrlAsHtml(commit.sha);
+  const type = commit.type === "Try" ? "Try" : "Master";
+  const pr = getArtifactPr(reason);
+
+  return `${type} commit ${url} (${pullRequestUrlAsHtml(pr)})`;
+}
+
+interface CurrentRun {
+  commit: Commit;
+  missing_reason: MissingReason;
+  start: Date;
+  expected_end: Date;
+  expected_total: number;
+  progress: number;
+}
+
+interface TimelineEntry {
+  pr: number | null;
+  kind: string;
+  sha: string;
+  status: string;
+  end_time: Date;
+  end_estimated: boolean;
+  duration: number | null;
+  errors: BenchmarkError[];
+  warning: boolean;
+  current: boolean;
 }
 
 const timeLeft = computed(() => {
   const steps = data.value?.current?.progress ?? [];
   let left = 0;
   for (const step of steps) {
     if (!step.is_done) {
-      left += step.expected_duration - step.current_progress;
+      left += Math.max(0, step.expected_duration - step.current_progress);
     }
   }
   return left;
 });
-const recentEndDate = computed(() => {
-  return new Date(data.value.most_recent_end * 1000);
+
+const currentRun: Ref<CurrentRun | null> = computed(() => {
+  const current = data.value?.current ?? null;
+  if (current === null) return null;
+  const sha = getArtifactSha(current.artifact);
+
+  const elapsed = current.progress.reduce(
+    (prev, current) => prev + current.current_progress,
+    0
+  );
+  const start = subSeconds(new Date(), elapsed);
+  const expected_total = current.progress.reduce(
+    (prev, current) => prev + current.expected_duration,
+    0
+  );
+  const progress = Math.min(1, elapsed / Math.max(1, expected_total));
+
+  for (const [commit, reason] of data.value.missing) {
+    if (commit.sha === sha && reason.hasOwnProperty("InProgress")) {
+      return {
+        commit,
+        missing_reason: reason["InProgress"],
+        start,
+        expected_end: addSeconds(new Date(), timeLeft.value),
+        expected_total,
+        progress,
+      };
+    }
+  }
+  return null;
+});
+const lastFinishedRun: Ref<FinishedRun | null> = computed(() => {
+  const finished = data?.value.finished_runs;
```

Comment:
it seems `value` or `finished_runs` can be null/undefined here

Reply:
Good point, this should have be `data.value?` instead.